### DISCUSSION TITLE
Loading relative line on operator pending wont load

### DISCUSSION THREAD

**Discussion by JoseConseco** [rasulomaroff/reactive.nvim#2] [Q&A] [answered]
Ok, its probably something simple but this:

```lua
  {
    'rasulomaroff/reactive.nvim', -- detect eg. operator pending mode
    config = function()
      require('reactive').setup{
        -- builtin = {
        --   cursorline = false,
        --   cursor = false,
        --   modemsg = false
        -- },
        configs = {
          -- a key here is a preset name, a value can be a boolean (if false, presets will be disabled)
          -- or a table that overwrites preset's values
          presetone = {
            name = 'test',
            modes = {
              no = {
                to = function()
                  vim.opt.relativenumber = true
                end,
                from = function()
                  vim.opt.relativenumber = false
                end,
              },
            }
          },
        }
      }
    end,
  },   
```
Does do nothing for me. I tried to call `Reactive enable presetone` but it says preset not found ....

**Comment by rasulomaroff** [comment, marked as answer]
Hi there! @JoseConseco
The reason nothing is happening because `configs` field is meant to configure already added presets.

To actually add a preset you have 2 options that listed in the end of the README file. I'm gonna change it so that it's easier for you to find it.

### **The first way**

1. You put your preset into `reactive/presets/yourpresetname.lua` folder & file
2. You then load it using `require('reactive').load_preset 'yourpresetname'`

### **The second way**

You use

```lua
require('reactive').add_preset {
  -- your config for preset here
}
```

I actually use the first way since it fits me better :)

Solution for your question:

```lua
require('reactive').add_preset {
  name = 'test',
  modes = {
    no = {
      to = function()
        vim.opt.relativenumber = true
      end,
      from = function()
        vim.opt.relativenumber = false
      end
    }
  }
}
```

You **don't** need to call `setup` function with this plugin :)

**Reply by JoseConseco** [reply]
Thanks it works now. Though  the  relative numbers only shows for brief moment.  What I tried to solve is : 
- lets say we want to delete   n lines ( with  dnj  - motion) 
-instead of counting  n lines down, i with to show relative nunmbers as soon as 'd' key is pressed. 
But right now the relative numbers are shown only  after  ` d + count + j ` keys are pressed.

see below
 - diw -  relative numbers are shows only for split second
 -  df(letter) - in that case we can see relative numbers afer  `df` keys combination (relative numbers do not help much in that case, but it works) 
 

https://github.com/rasulomaroff/reactive.nvim/assets/13521338/090f45b8-491c-4d19-b822-4b7aaf81b44c


but if lets say I try to delete 22 lines,  then  d22j - will show relative numbers at the very end of  entering the sequence.

**Reply by rasulomaroff** [reply]
Okay, and what will happen if you press 'd' and wait for a moment? Will it show relative numbers eventually? If so, then the reason behind this is that your config uses 'd' for something else as well and neovim waits for a moment to see whether you're gonna continue combination. Often it happens because of plugins as 'mini.surround' or  'nvim-surround'. They have combinations like `ds'` which don't evoke operator-pending mode. And that's why neovim is waiting for you to continue your combination.

You can ask why it shows relative numbers after you press `d` + `count`. Well, in this case:
1. You have 2 'collisional' mappings - `d{count}{where}` and `ds'`
2. When you press `d` neovim just doesn't know what's going to happen next since `ds'` isn't an operator pending mapping.
3. Then, when you press `count` for example, neovim understands that there's only delete operator left and it performs mode switching, thus executing those callbacks from your config.

Can you please tell me whether your case refers to what I've just said?

**Comment by JoseConseco** [comment]
@rasulomaroff It works. Indeed that was collision mapping - removing which key  addon made it work. Now I press just `d` - can look on relative line number I want to delete to, then finish with  `xj`  (x - count) 

https://github.com/rasulomaroff/reactive.nvim/assets/13521338/c3521898-8ed4-4fc0-94f6-4dc72cd60e11

The only issues is , I cannot live without  https://github.com/folke/which-key.nvim :/  . Is tere any chance for you to support  which-key. nvim  ? 
Anyway, awesome addon.  I was hoping for this kind of workflow being possible for long time!

**Reply by rasulomaroff** [reply]
Great that we found it out! I understand your frustration about the compatibility with `which-key`, but I can't actually see anything I can do in this case since I'm just watching for the `ModeChanged` event to happen and react to it.

The only solution I see right now is just to disable `which-key` for those shortcuts (you can do it in its config), but I don't think it would work for you

**Reply by JoseConseco** [reply]
I was just thinking, since you are watching `ModeChange` event - it should not be the case that other plugin (which-key) in intersecting this event, and 'muting' it?  I mean if this is how it works, then it would mean, only one plugin can listen to give event.   But it is not how it works - multiple plugins should be able to listen for any event right?    Or is it that you try to override `d` key, but it is already overriden by another plugin, and that is why reactive wont work with which key?

**Reply by rasulomaroff** [reply]
When I checked how `which-key` works, it was working like that:

1. Takes control of user input
2. After delay shows available shortcuts and fires that input
3. In case you pressing next keys without waiting for that delay - fires input without waiting

The main thing to know here is that it takes control over what's firing to Neovim's input. It worked this way some time ago and I think it does this right now as well.
That's why you don't get your `d` pressed right away. You can add `ModeChanged` autocommand yourself and check when it's actually executed after pressing `d` for example.

You understood it exactly right - every plugin should be able to listen for any event without intersecting with each other and it's like that now, but as I said, you can do nothing when something just takes control over your input or has collided keymaps as in other cases we discussed, where Neovim itself waits for some time to understand what you're intending to do.

**Reply by JoseConseco** [reply]
ok, so if 1. is true - then I assume it must be blocking function. and it block input for lets say 200ms, and this makes it impossible for  reactive to do anything in nvim right? If that delay was nonblocking,  then I think you could execute your code in that window ? I saw folke did not do any activity  for 3months now, so I have low hopes, he could change his code ...

**Reply by rasulomaroff** [reply]
Basically, he won't be able to achieve the behaviour he wants other way. So, it's up to everybody to decide whether they need those plugins or not.

Regarding your question: I can't really say how his plugin is organized because I don't really remember (I checked it half a year ago), but I think it's done in a very smart way and if there was a chance to do that other (better) way, I believe he would do that. Because not only `reactive.nvim` has these types of troubles with `which-key`, but other plugins either.

As I already said, I don't monitor user input nor delay anything. The only thing `reactive.nvim` does is just listening to those events and executes callbacks/applies highlights when they're triggered.

**Comment by FedeAbella** [comment]
This was SO helpful! I realised early on `which-key` was messing with some of my plugins, but even after removing it couldn't figure for the life of me why reactive was not working on operators. Turns out nvim was waiting for `nvim-surround`.

**Reply by rasulomaroff** [reply]
Haha so true, I experienced that a lot in the past as well.

I also use `nvim-surround` right now and even though I changed its default mappings to be another ones, it still has those `dS`, `dss` mappings. So, will need to change those as well

**Comment by JoseConseco** [comment]
Just wanted to say, that new Which key is out , and now reactive works ok with it. Soo cool.. 
My config that makes relative number on operator pending mode; 
```lua
  {
    'rasulomaroff/reactive.nvim', -- detect eg. operator pending mode
    config = function()
      require('reactive').setup{
        -- builtin = {
        --   cursorline = false,
        --   cursor = false,
        --   modemsg = false
        -- },
        configs = {
          -- a key here is a preset name, a value can be a boolean (if false, presets will be disabled)
          -- or a table that overwrites preset's values
        }
      }
      require('reactive').add_preset {
        name = 'test',
        modes = {
          no = {
            to = function()
              vim.opt.relativenumber = true
            end,
            from = function()
              vim.opt.relativenumber = false
            end
          }
        }
      }
    end,
  },   
```

**Reply by rasulomaroff** [reply]
Great! Nice to hear that!
